File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/move/StockPileToFoundationMove.java
package org.silnith.game.solitaire.move;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import org.silnith.deck.Card;
import org.silnith.deck.Suit;
import org.silnith.game.solitaire.Board;
import org.silnith.game.util.Pair;


/**
 * A move that takes a card from the stock pile and puts it into the Foundation.
 */
public class StockPileToFoundationMove implements SolitaireMove {
	
	/**
	 * Finds all moves where a card is drawn from the stock pile to the foundation.
	 * 
	 * <p>This will either contain one move or zero.</p>
	 * 
	 * @param board the board to examine
	 * @return a collection of moves
	 */
	public static Collection<StockPileToFoundationMove> findMoves(final Board board) {
		if (board.getStockPileIndex() > 0) {
			final Card card = board.getStockPileCard();
			if (board.canAddToFoundation(card)) {
				return Collections.singleton(new StockPileToFoundationMove(board));
			}
		}
		return Collections.emptySet();
	}
    
    /**
     * The index into the stock pile from which the card is being taken.
     */
    private final int sourceIndex;
    
    /**
     * The card being moved.
     */
    private final Card card;
    
    /**
     * Creates a new move that takes a card from the stock pile and puts it into
     * the Foundation.
     * 
     * @param sourceIndex the index into the stock pile of the card being moved
     * @param card the card being moved
     */
    public StockPileToFoundationMove(final int sourceIndex, final Card card) {
        super();
        this.sourceIndex = sourceIndex;
        this.card = card;
    }
    
    /**
     * Creates a new move that takes a card from the stock pile and puts it into
     * the foundation.
     * 
     * <p>This takes the parameters from the current state of the provided board.</p>
     * 
     * @param board the board from which to get the source index and card
     * @throws IndexOutOfBoundsException if the board stock pile index is
     *         {@code 0} or greater than the size of the stock pile
     */
    public StockPileToFoundationMove(final Board board) {
    	this(board.getStockPileIndex(), board.getStockPileCard());
    }
    
    /**
     * Returns the index into the stock pile of the card being moved.
     * 
     * @return the stock pile index
     */
    public int getSourceIndex() {
        return sourceIndex;
    }
    
    /**
     * Returns the card being moved.
     * 
     * @return the card being moved
     * @see #getCards()
     */
    public Card getCard() {
        return card;
    }
    
    @Override
    public boolean hasCards() {
        return true;
    }
    
    @Override
    public List<Card> getCards() {
        return Collections.singletonList(card);
    }

    @Override
    public boolean isStockPileModification() {
        return true;
    }

    @Override
    public boolean isStockPileAdvance() {
        return false;
    }

    @Override
    public boolean isStockPileRecycle() {
        return false;
    }

    @Override
    public boolean isFromStockPile() {
        return true;
    }

    @Override
    public boolean isFromFoundation() {
        return false;
    }

    @Override
    public boolean isFromColumn() {
        return false;
    }

    @Override
    public boolean isFromColumn(int columnIndex) {
        return false;
    }

    @Override
    public int getSourceColumnIndex() {
        throw new IllegalStateException("Not a move from a column.");
    }

    @Override
    public boolean isToFoundation() {
        return true;
    }

    @Override
    public boolean isToColumn() {
        return false;
    }

    @Override
    public boolean isToColumn(int columnIndex) {
        return false;
    }

    @Override
    public int getDestinationColumnIndex() {
        throw new IllegalStateException("Not a move to a column.");
    }

	@Override
    public Board apply(final Board board) {
        final Pair<Card, List<Card>> pair = board.extractStockPileCard();
		final Card card = pair.getFirst();
		final List<Card> newStockPile = pair.getSecond();
		
		final int newStockPileIndex = board.getStockPileIndex() - 1;
		
		final Map<Suit, List<Card>> newFoundation = board.getFoundationPlusCard(card);
		
		return new Board(board.getColumns(), newStockPile, newStockPileIndex, newFoundation);
    }
    
    @Override
    public int hashCode() {
        return sourceIndex ^ card.hashCode();
    }
    
    @Override
    public boolean equals(final Object obj) {
        if (obj instanceof StockPileToFoundationMove) {
            final StockPileToFoundationMove move = (StockPileToFoundationMove) obj;
            return sourceIndex == move.sourceIndex && card.equals(move.card);
        } else {
            return false;
        }
    }
    
    @Override
    public String toString() {
        return "Move " + card + " from stock pile " + sourceIndex + " to foundation.";
    }
    
}

File: cpp/Solitaire/Klondike/silnith/game/solitaire/move/StockPileToFoundationMove.cpp
#include <silnith/game/solitaire/move/StockPileToFoundationMove.h>

#include <silnith/game/deck/card.h>

using namespace silnith::game::deck;

using namespace std;
using namespace std::literals::string_literals;

namespace silnith::game::solitaire::move
{
    vector<shared_ptr<solitaire_move>> StockPileToFoundationMove::find_moves(board const& board)
    {
        if (board.get_stock_pile_index() > 0)
        {
            if (board.can_add_to_foundation(board.get_stock_pile_card()))
            {
                return vector<shared_ptr<solitaire_move>>{ make_shared<StockPileToFoundationMove>(board) };
            }
        }
        return vector<shared_ptr<solitaire_move>>{};
    }

    StockPileToFoundationMove::StockPileToFoundationMove(size_t source_index, card const& card)
        : source_index{ source_index },
        _card{ card }
    {}

    StockPileToFoundationMove::StockPileToFoundationMove(board const& board)
        : StockPileToFoundationMove(board.get_stock_pile_index(), board.get_stock_pile_card())
    {}

    StockPileToFoundationMove::StockPileToFoundationMove(shared_ptr<board> const& board)
        : StockPileToFoundationMove(*board)
    {}

    size_t StockPileToFoundationMove::get_source_index(void) const
    {
        return source_index;
    }

    bool StockPileToFoundationMove::has_cards(void) const
    {
        return true;
    }

    vector<card> StockPileToFoundationMove::get_cards(void) const
    {
        return vector<card>{ _card };
    }

    bool StockPileToFoundationMove::is_stock_pile_modification(void) const
    {
        return true;
    }

    bool StockPileToFoundationMove::is_stock_pile_advance(void) const
    {
        return false;
    }

    bool StockPileToFoundationMove::is_stock_pile_recycle(void) const
    {
        return false;
    }

    bool StockPileToFoundationMove::is_from_stock_pile(void) const
    {
        return true;
    }

    bool StockPileToFoundationMove::is_from_foundation(void) const
    {
        return false;
    }

    bool StockPileToFoundationMove::is_from_column(void) const
    {
        return false;
    }

    bool StockPileToFoundationMove::is_from_column(std::size_t column_index) const
    {
        return false;
    }

    size_t StockPileToFoundationMove::get_source_column_index(void) const
    {
        throw invalid_argument("Not a move from a column.");
    }

    bool StockPileToFoundationMove::is_to_foundation(void) const
    {
        return true;
    }

    bool StockPileToFoundationMove::is_to_column(void) const
    {
        return false;
    }

    bool StockPileToFoundationMove::is_to_column(size_t column_index) const
    {
        return false;
    }

    size_t StockPileToFoundationMove::get_destination_column_index(void) const
    {
        throw invalid_argument("Not a move to a column.");
    }

    shared_ptr<board const> StockPileToFoundationMove::apply(shared_ptr<board const> const& b) const
    {
        pair<card, vector<card>> pair{ b->extract_stock_pile_card() };
        card drawn_card{ pair.first };
        vector<card> new_stock_pile{ pair.second };

        size_t stock_pile_index{ b->get_stock_pile_index() };
        size_t new_stock_pile_index{ stock_pile_index - 1 };

        map<suit, vector<card>> new_foundation{ b->get_foundation_plus_card(drawn_card) };

        return make_shared<board>(
            b->get_columns(),
            new_stock_pile,
            new_stock_pile_index,
            new_foundation);
    }

    bool StockPileToFoundationMove::operator==(StockPileToFoundationMove const& other) const
    {
        return source_index == other.source_index
            && _card == other._card;
    }

    string to_string(StockPileToFoundationMove const& move)
    {
        return "Move "s
            + to_string(move.get_cards().front())
            + " from stock pile index "s
            + std::to_string(move.get_source_index())
            + " to foundation."s;
    }

    wstring to_wstring(StockPileToFoundationMove const& move)
    {
        return L"Move "s
            + to_wstring(move.get_cards().front())
            + L" from stock pile index "s
            + std::to_wstring(move.get_source_index())
            + L" to foundation."s;
    }

    ostream& operator<<(ostream& out, StockPileToFoundationMove const& move)
    {
        out << to_string(move);
        return out;
    }

    wostream& operator<<(wostream& out, StockPileToFoundationMove const& move)
    {
        out << to_wstring(move);
        return out;
    }
}
